File: airbyte-integrations/connectors/destination-mysql/src/test-integration/java/io/airbyte/integrations/destination/mysql/typing_deduping/MysqlSqlGeneratorIntegrationTest.java
 /*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.mysql.typing_deduping;

import static io.airbyte.cdk.integrations.base.JavaBaseConstants.COLUMN_NAME_AB_EXTRACTED_AT;
import static io.airbyte.cdk.integrations.base.JavaBaseConstants.COLUMN_NAME_AB_ID;
import static io.airbyte.cdk.integrations.base.JavaBaseConstants.COLUMN_NAME_AB_LOADED_AT;
import static io.airbyte.cdk.integrations.base.JavaBaseConstants.COLUMN_NAME_AB_RAW_ID;
import static io.airbyte.cdk.integrations.base.JavaBaseConstants.COLUMN_NAME_DATA;
import static io.airbyte.cdk.integrations.base.JavaBaseConstants.COLUMN_NAME_EMITTED_AT;
import static io.airbyte.integrations.destination.mysql.typing_deduping.MysqlSqlGenerator.TIMESTAMP_FORMATTER;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.airbyte.cdk.db.jdbc.DefaultJdbcDatabase;
import io.airbyte.cdk.db.jdbc.JdbcDatabase;
import io.airbyte.cdk.db.jdbc.JdbcSourceOperations;
import io.airbyte.cdk.integrations.destination.jdbc.TableDefinition;
import io.airbyte.cdk.integrations.destination.jdbc.typing_deduping.JdbcDestinationHandler;
import io.airbyte.cdk.integrations.destination.jdbc.typing_deduping.JdbcSqlGenerator;
import io.airbyte.cdk.integrations.standardtest.destination.typing_deduping.JdbcSqlGeneratorIntegrationTest;
import io.airbyte.commons.json.Jsons;
import io.airbyte.integrations.base.destination.typing_deduping.DestinationHandler;
import io.airbyte.integrations.base.destination.typing_deduping.Sql;
import io.airbyte.integrations.base.destination.typing_deduping.StreamId;
import io.airbyte.integrations.destination.mysql.MySQLDestination;
import io.airbyte.integrations.destination.mysql.MySQLDestinationAcceptanceTest;
import io.airbyte.integrations.destination.mysql.MySQLNameTransformer;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Optional;
import javax.sql.DataSource;
import org.jooq.DataType;
import org.jooq.Field;
import org.jooq.SQLDialect;
import org.jooq.conf.ParamType;
import org.jooq.impl.DSL;
import org.jooq.impl.DefaultDataType;
import org.jooq.impl.SQLDataType;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.testcontainers.containers.MySQLContainer;

public class MysqlSqlGeneratorIntegrationTest extends JdbcSqlGeneratorIntegrationTest {

  private static MySQLContainer<?> testContainer;
  private static JdbcDatabase database;

  public static class MysqlSourceOperations extends JdbcSourceOperations {

    @Override
    public void copyToJsonField(final ResultSet resultSet, final int colIndex, final ObjectNode json) throws SQLException {
      final String columnName = resultSet.getMetaData().getColumnName(colIndex);
      final String columnTypeName = resultSet.getMetaData().getColumnTypeName(colIndex).toLowerCase();

      // JSONB has no equivalent in JDBCType
      if ("json".equals(columnTypeName)) {
        json.set(columnName, Jsons.deserializeExact(resultSet.getString(colIndex)));
      } else {
        super.copyToJsonField(resultSet, colIndex, json);
      }
    }

  }

  @BeforeAll
  public static void setupMysql() throws Exception {
    testContainer = new MySQLContainer<>("mysql:8.0");
    testContainer.start();
    MySQLDestinationAcceptanceTest.configureTestContainer(testContainer);

    final JsonNode config = MySQLDestinationAcceptanceTest.getConfigFromTestContainer(testContainer);

    // TODO move this into JdbcSqlGeneratorIntegrationTest?
    // This code was largely copied from RedshiftSqlGeneratorIntegrationTest
    // TODO: Its sad to instantiate unneeded dependency to construct database and datsources. pull it to
    // static methods.
    final MySQLDestination insertDestination = new MySQLDestination();
    final DataSource dataSource = insertDestination.getDataSource(config);
    database = new DefaultJdbcDatabase(dataSource, new MysqlSourceOperations());
  }

  @AfterAll
  public static void teardownMysql() {
    testContainer.stop();
    testContainer.close();
  }

  @Override
  protected JdbcSqlGenerator getSqlGenerator() {
    return new MysqlSqlGenerator(new MySQLNameTransformer());
  }

  @Override
  protected DestinationHandler<TableDefinition> getDestinationHandler() {
    // Mysql doesn't have an actual schema concept.
    // All of our queries pass a value into the "schemaName" parameter, which mysql treats as being
    // the database name.
    // So we pass null for the databaseName parameter here, because we don't use the 'test' database at all.
    return new JdbcDestinationHandler(null, database, SQLDialect.MYSQL);
  }

  @Override
  protected void insertRawTableRecords(final StreamId streamId, final List<JsonNode> records) throws Exception {
    reformatMetaColumnTimestamps(records);
    super.insertRawTableRecords(streamId, records);
  }

  @Override
  protected void insertFinalTableRecords(final boolean includeCdcDeletedAt,
                                         final StreamId streamId,
                                         final String suffix,
                                         final List<JsonNode> records)
      throws Exception {
    reformatMetaColumnTimestamps(records);
    super.insertFinalTableRecords(includeCdcDeletedAt, streamId, suffix, records);
  }

  @Override
  protected void insertV1RawTableRecords(final StreamId streamId, final List<JsonNode> records) throws Exception {
    reformatMetaColumnTimestamps(records);
    super.insertV1RawTableRecords(streamId, records);
  }

  private static void reformatMetaColumnTimestamps(final List<JsonNode> records) {
    // We use mysql's TIMESTAMP(6) type for extracted_at+loaded_at.
    // Unfortunately, mysql doesn't allow you to use the 'Z' suffix for UTC timestamps.
    // Convert those to '+00:00' here.
    for (final JsonNode record : records) {
      reformatTimestampIfPresent(record, COLUMN_NAME_AB_EXTRACTED_AT);
      reformatTimestampIfPresent(record, COLUMN_NAME_EMITTED_AT);
      reformatTimestampIfPresent(record, COLUMN_NAME_AB_LOADED_AT);
    }
  }

  private static void reformatTimestampIfPresent(final JsonNode record, final String columnNameAbExtractedAt) {
    if (record.has(columnNameAbExtractedAt)) {
      final OffsetDateTime extractedAt = OffsetDateTime.parse(record.get(columnNameAbExtractedAt).asText());
      final String reformattedExtractedAt = TIMESTAMP_FORMATTER.format(extractedAt);
      ((ObjectNode) record).put(columnNameAbExtractedAt, reformattedExtractedAt);
    }
  }

  @Override
  protected void createRawTable(final StreamId streamId) throws Exception {
    getDatabase().execute(getDslContext().createTable(DSL.name(streamId.rawNamespace(), streamId.rawName()))
        .column(COLUMN_NAME_AB_RAW_ID, SQLDataType.VARCHAR(36).nullable(false))
        // we use VARCHAR for timestamp values, but TIMESTAMP(6) for extracted+loaded_at.
        // because legacy normalization did that. :shrug:
        .column(COLUMN_NAME_AB_EXTRACTED_AT, SQLDataType.TIMESTAMP(6).nullable(false))
        .column(COLUMN_NAME_AB_LOADED_AT, SQLDataType.TIMESTAMP(6))
        .column(COLUMN_NAME_DATA, getStructType().nullable(false))
        .getSQL(ParamType.INLINED));
  }

  @Override
  protected void createV1RawTable(final StreamId v1RawTable) throws Exception {
    getDatabase().execute(getDslContext().createTable(DSL.name(v1RawTable.rawNamespace(), v1RawTable.rawName()))
        .column(COLUMN_NAME_AB_ID, SQLDataType.VARCHAR(36).nullable(false))
        // similar to createRawTable - this data type is timestmap, not varchar
        .column(COLUMN_NAME_EMITTED_AT, SQLDataType.TIMESTAMP(6).nullable(false))
        .column(COLUMN_NAME_DATA, getStructType().nullable(false))
        .getSQL(ParamType.INLINED));
  }

  @Test
  @Override
  public void testCreateTableIncremental() throws Exception {
    // TODO
    final Sql sql = generator.createTable(incrementalDedupStream, "", false);
    destinationHandler.execute(sql);

    final Optional<TableDefinition> existingTable = destinationHandler.findExistingTable(incrementalDedupStream.id());

    assertTrue(existingTable.isPresent());
    assertAll(
        () -> assertEquals("VARCHAR", existingTable.get().columns().get("_airbyte_raw_id").type()),
        () -> assertEquals("TIMESTAMP", existingTable.get().columns().get("_airbyte_extracted_at").type()),
        () -> assertEquals("JSON", existingTable.get().columns().get("_airbyte_meta").type()),
        () -> assertEquals("BIGINT", existingTable.get().columns().get("id1").type()),
        () -> assertEquals("BIGINT", existingTable.get().columns().get("id2").type()),
        () -> assertEquals("VARCHAR", existingTable.get().columns().get("updated_at").type()),
        () -> assertEquals("JSON", existingTable.get().columns().get("struct").type()),
        () -> assertEquals("JSON", existingTable.get().columns().get("array").type()),
        // Note that we use text here, since mysql varchar is limited to 64K
        () -> assertEquals("TEXT", existingTable.get().columns().get("string").type()),
        () -> assertEquals("DECIMAL", existingTable.get().columns().get("number").type()),
        () -> assertEquals("BIGINT", existingTable.get().columns().get("integer").type()),
        () -> assertEquals("BIT", existingTable.get().columns().get("boolean").type()),
        () -> assertEquals("VARCHAR", existingTable.get().columns().get("timestamp_with_timezone").type()),
        () -> assertEquals("DATETIME", existingTable.get().columns().get("timestamp_without_timezone").type()),
        () -> assertEquals("VARCHAR", existingTable.get().columns().get("time_with_timezone").type()),
        () -> assertEquals("TIME", existingTable.get().columns().get("time_without_timezone").type()),
        () -> assertEquals("DATE", existingTable.get().columns().get("date").type()),
        () -> assertEquals("JSON", existingTable.get().columns().get("unknown").type()));
    // TODO assert on table clustering, etc.
  }

  @Override
  protected JdbcDatabase getDatabase() {
    return database;
  }

  @Override
  protected DataType<?> getStructType() {
    return new DefaultDataType<>(null, String.class, "json");
  }

  @Override
  protected SQLDialect getSqlDialect() {
    return SQLDialect.MYSQL;
  }

  @Override
  protected Field<?> toJsonValue(final String valueAsString) {
    // mysql lets you just insert json strings directly into json columns
    return DSL.val(valueAsString);
  }

  @Override
  protected void teardownNamespace(final String namespace) throws Exception {
    // mysql doesn't have a CASCADE keyword in DROP SCHEMA, so we have to override this method.
    // we're currently on jooq 3.13; jooq's dropDatabase() call was only added in 3.14
    getDatabase().execute(getDslContext().dropSchema(namespace).getSQL(ParamType.INLINED));
  }

  @Override
  public boolean supportsSafeCast() {
    return false;
  }

}

File: airbyte-integrations/connectors/destination-mysql/src/test-integration/kotlin/io/airbyte/integrations/destination/mysql/typing_deduping/MysqlSqlGeneratorIntegrationTest.kt
/*
 * Copyright (c) 2026 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.integrations.destination.mysql.typing_deduping

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ObjectNode
import io.airbyte.cdk.db.jdbc.DefaultJdbcDatabase
import io.airbyte.cdk.db.jdbc.JdbcDatabase
import io.airbyte.cdk.integrations.base.JavaBaseConstants.COLUMN_NAME_AB_EXTRACTED_AT
import io.airbyte.cdk.integrations.base.JavaBaseConstants.COLUMN_NAME_AB_ID
import io.airbyte.cdk.integrations.base.JavaBaseConstants.COLUMN_NAME_AB_LOADED_AT
import io.airbyte.cdk.integrations.base.JavaBaseConstants.COLUMN_NAME_AB_META
import io.airbyte.cdk.integrations.base.JavaBaseConstants.COLUMN_NAME_AB_RAW_ID
import io.airbyte.cdk.integrations.base.JavaBaseConstants.COLUMN_NAME_DATA
import io.airbyte.cdk.integrations.base.JavaBaseConstants.COLUMN_NAME_EMITTED_AT
import io.airbyte.cdk.integrations.destination.jdbc.typing_deduping.JdbcSqlGenerator
import io.airbyte.cdk.integrations.standardtest.destination.typing_deduping.JdbcSqlGeneratorIntegrationTest
import io.airbyte.commons.json.Jsons
import io.airbyte.integrations.base.destination.typing_deduping.DestinationHandler
import io.airbyte.integrations.base.destination.typing_deduping.StreamId
import io.airbyte.integrations.base.destination.typing_deduping.migrators.MinimumDestinationState
import io.airbyte.integrations.destination.mysql.MySQLDestination
import io.airbyte.integrations.destination.mysql.MySQLNameTransformer
import io.airbyte.integrations.destination.mysql.MysqlTestDatabase
import io.airbyte.integrations.destination.mysql.MysqlTestSourceOperations
import io.airbyte.integrations.destination.mysql.typing_deduping.MysqlSqlGenerator.Companion.TIMESTAMP_FORMATTER
import java.time.OffsetDateTime
import java.util.Optional
import org.jooq.DataType
import org.jooq.Field
import org.jooq.SQLDialect
import org.jooq.conf.ParamType
import org.jooq.impl.DSL
import org.jooq.impl.SQLDataType
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test

class MysqlSqlGeneratorIntegrationTest :
    JdbcSqlGeneratorIntegrationTest<MinimumDestinationState>() {

    override val sqlDialect: SQLDialect = SQLDialect.MYSQL
    override val sqlGenerator: JdbcSqlGenerator = MysqlSqlGenerator()
    override val structType: DataType<*> = MysqlSqlGenerator.JSON_TYPE
    override val supportsSafeCast: Boolean = false
    override val database = Companion.database
    override val destinationHandler: DestinationHandler<MinimumDestinationState>
        // lazy init. We need `namespace` to be initialized before this call.
        get() = MysqlDestinationHandler(Companion.database, namespace)

    @Throws(Exception::class)
    override fun insertRawTableRecords(streamId: StreamId, records: List<JsonNode>) {
        reformatMetaColumnTimestamps(records)
        super.insertRawTableRecords(streamId, records)
    }

    @Throws(Exception::class)
    override fun insertFinalTableRecords(
        includeCdcDeletedAt: Boolean,
        streamId: StreamId,
        suffix: String?,
        records: List<JsonNode>
    ) {
        reformatMetaColumnTimestamps(records)
        super.insertFinalTableRecords(includeCdcDeletedAt, streamId, suffix, records)
    }

    @Throws(Exception::class)
    override fun insertV1RawTableRecords(streamId: StreamId, records: List<JsonNode>) {
        reformatMetaColumnTimestamps(records)
        super.insertV1RawTableRecords(streamId, records)
    }

    @Throws(Exception::class)
    override fun createRawTable(streamId: StreamId) {
        database.execute(
            dslContext
                .createTable(DSL.name(streamId.rawNamespace, streamId.rawName))
                .column(COLUMN_NAME_AB_RAW_ID, SQLDataType.VARCHAR(256).nullable(false))
                .column(COLUMN_NAME_DATA, structType.nullable(false))
                // we use VARCHAR for timestamp values, but TIMESTAMP(6) for extracted+loaded_at.
                // because legacy normalization did that. :shrug:
                .column(COLUMN_NAME_AB_EXTRACTED_AT, SQLDataType.TIMESTAMP(6).nullable(false))
                .column(COLUMN_NAME_AB_LOADED_AT, SQLDataType.TIMESTAMP(6))
                .column(COLUMN_NAME_AB_META, structType.nullable(true))
                .getSQL(ParamType.INLINED),
        )
    }

    @Throws(Exception::class)
    override fun createV1RawTable(v1RawTable: StreamId) {
        database.execute(
            dslContext
                .createTable(DSL.name(v1RawTable.rawNamespace, v1RawTable.rawName))
                .column(
                    COLUMN_NAME_AB_ID,
                    SQLDataType.VARCHAR(36).nullable(false),
                ) // similar to createRawTable - this data type is timestmap, not varchar
                .column(COLUMN_NAME_EMITTED_AT, SQLDataType.TIMESTAMP(6).nullable(false))
                .column(COLUMN_NAME_DATA, structType.nullable(false))
                .getSQL(ParamType.INLINED),
        )
    }

    @Test
    @Throws(Exception::class)
    override fun testCreateTableIncremental() {
        val sql = generator.createTable(incrementalDedupStream, "", false)
        destinationHandler.execute(sql)

        val initialStatuses = destinationHandler.gatherInitialState(listOf(incrementalDedupStream))
        Assertions.assertEquals(1, initialStatuses.size)
        val initialStatus = initialStatuses.first()
        Assertions.assertTrue(initialStatus.isFinalTablePresent)
        Assertions.assertFalse(initialStatus.isSchemaMismatch)
    }

    override fun toJsonValue(valueAsString: String?): Field<*> {
        // mysql lets you just insert json strings directly into json columns
        return DSL.`val`(valueAsString)
    }

    override fun createNamespace(namespace: String) {
        database.execute(
            dslContext
                .createSchemaIfNotExists(nameTransformer.getIdentifier(namespace))
                .getSQL(ParamType.INLINED)
        )
    }

    override fun teardownNamespace(namespace: String) {
        database.execute(
            dslContext
                .dropDatabaseIfExists(nameTransformer.getIdentifier(namespace))
                .getSQL(ParamType.INLINED)
        )
    }

    /**
     * Verify that strings longer than 512 characters survive typing and deduping. MySQL's
     * JSON_VALUE() defaults to VARCHAR(512) and silently returns NULL for longer values. See:
     * https://github.com/airbytehq/airbyte/issues/73303
     */
    @Test
    @Throws(Exception::class)
    fun testLongStringNotTruncated() {
        val longString = "a".repeat(600)

        createRawTable(streamId)
        createFinalTable(incrementalAppendStream, "")
        insertRawTableRecords(
            streamId,
            listOf(
                Jsons.deserialize(
                    """
                    {
                        "_airbyte_raw_id": "ad3e8a20-5328-4b5f-8841-1a1264e5bfc0",
                        "_airbyte_extracted_at": "2023-01-01T00:00:00Z",
                        "_airbyte_data": {
                            "id1": 1,
                            "id2": 100,
                            "updated_at": "2023-01-01T01:00:00Z",
                            "string": "$longString"
                        }
                    }
                    """
                )
            ),
        )

        val sql = generator.updateTable(incrementalAppendStream, "", Optional.empty(), false)
        destinationHandler.execute(sql)

        val finalRecords = dumpFinalTableRecords(streamId, "")
        Assertions.assertEquals(1, finalRecords.size, "Expected exactly one final record")
        val stringValue = finalRecords[0].get("string")
        Assertions.assertNotNull(
            stringValue,
            "String field should not be null after typing/deduping"
        )
        Assertions.assertFalse(
            stringValue.isNull,
            "String field should not be JSON null after typing/deduping"
        )
        Assertions.assertEquals(
            longString,
            stringValue.asText(),
            "600-char string should survive typing/deduping without truncation"
        )
    }

    companion object {
        private lateinit var testContainer: MysqlTestDatabase
        private lateinit var database: JdbcDatabase
        private val nameTransformer = MySQLNameTransformer()

        @JvmStatic
        @BeforeAll
        @Throws(Exception::class)
        fun setupMysql() {
            testContainer = MysqlTestDatabase.`in`(MysqlTestDatabase.BaseImage.MYSQL_8)

            val config =
                testContainer
                    .configBuilder()
                    .withDatabase()
                    .withHostAndPort()
                    .withCredentials()
                    .withoutSsl()
                    .build()

            // TODO move this into JdbcSqlGeneratorIntegrationTest?
            // This code was largely copied from RedshiftSqlGeneratorIntegrationTest
            // TODO: Its sad to instantiate unneeded dependency to construct database and
            // datsources. pull it to
            // static methods.
            database =
                DefaultJdbcDatabase(
                    MySQLDestination().getDataSource(config),
                    MysqlTestSourceOperations(),
                )
        }

        @JvmStatic
        @AfterAll
        fun teardownMysql() {
            // Intentionally do nothing.
            // The testcontainer will die at the end of the test run.
        }

        private fun reformatMetaColumnTimestamps(records: List<JsonNode>) {
            // We use mysql's TIMESTAMP(6) type for extracted_at+loaded_at.
            // Unfortunately, mysql doesn't allow you to use the 'Z' suffix for UTC timestamps.
            // Convert those to '+00:00' here.
            for (record in records) {
                reformatTimestampIfPresent(record, COLUMN_NAME_AB_EXTRACTED_AT)
                reformatTimestampIfPresent(record, COLUMN_NAME_EMITTED_AT)
                reformatTimestampIfPresent(record, COLUMN_NAME_AB_LOADED_AT)
            }
        }

        private fun reformatTimestampIfPresent(record: JsonNode, columnNameAbExtractedAt: String) {
            if (record.has(columnNameAbExtractedAt)) {
                val extractedAt = OffsetDateTime.parse(record[columnNameAbExtractedAt].asText())
                val reformattedExtractedAt: String = TIMESTAMP_FORMATTER.format(extractedAt)
                (record as ObjectNode).put(columnNameAbExtractedAt, reformattedExtractedAt)
            }
        }
    }
}
